File: android/app/src/main/java/com/hermes/client/data/network/SearchQuery.kt
package com.hermes.client.data.network

/**
 * Shapes the user's raw search text for the gateway's `/api/sessions/search`.
 *
 * The gateway appends a prefix wildcard to every unquoted token ("nimb" → "nimb*"). That is
 * right for Latin words but breaks CJK on the production store: the CJK path (trigram index, or
 * the LIKE fallback for 1–2 char terms) treats `词*` as a literal that never matches (verified
 * against the Mac mini on 2026-09-03: `的历史记录` → 1 hit, `的历史记录*` → 0). Quoting a token
 * makes the gateway keep it as-is, and a quoted phrase is exact substring matching on the CJK
 * paths — which is what a Chinese reader expects. Latin tokens stay unquoted so partial words
 * keep matching.
 */
fun buildSearchQuery(raw: String): String {
    val tokens = TOKEN.findAll(raw.trim()).map { it.value }.toList()
    return tokens.joinToString(" ") { token ->
        when {
            token.startsWith("\"") -> token
            token.endsWith("*") -> token
            token.any { it.isCjk() } -> "\"" + token.replace("\"", "") + "\""
            else -> token
        }
    }
}

/** True for CJK unified ideographs, kana, and Hangul — scripts the gateway's wildcard breaks. */
fun Char.isCjk(): Boolean {
    val c = code
    return c in 0x4E00..0x9FFF || // CJK Unified Ideographs
        c in 0x3400..0x4DBF ||    // Extension A
        c in 0xF900..0xFAFF ||    // Compatibility Ideographs
        c in 0x3040..0x30FF ||    // Hiragana, Katakana
        c in 0xAC00..0xD7AF       // Hangul syllables
}

/** True when [text] contains any CJK character. */
fun containsCjk(text: String): Boolean = text.any { it.isCjk() }

private val TOKEN = Regex("\"[^\"]*\"|\\S+")

File: web/src/hermes/search-query.ts
// Port of android data/network/SearchQuery.kt. The Hermes search appends a prefix wildcard to every
// unquoted token, which breaks CJK (`的历史记录*` → 0 hits), so CJK tokens are quoted; Latin tokens
// stay unquoted so partial words keep matching.

/** Java `\S` is ASCII-only: an ideographic space (U+3000) is part of a token, as on Android. */
const TOKEN = /"[^"]*"|[^ \t\n\x0B\f\r]+/g;

/** CJK unified ideographs (+Ext A, compatibility), kana and Hangul syllables. Per UTF-16 unit. */
export function isCjk(code: number): boolean {
  return (
    (code >= 0x4e00 && code <= 0x9fff) ||
    (code >= 0x3400 && code <= 0x4dbf) ||
    (code >= 0xf900 && code <= 0xfaff) ||
    (code >= 0x3040 && code <= 0x30ff) ||
    (code >= 0xac00 && code <= 0xd7af)
  );
}

export function containsCjk(text: string): boolean {
  for (let i = 0; i < text.length; i++) if (isCjk(text.charCodeAt(i))) return true;
  return false;
}

export function buildSearchQuery(raw: string): string {
  // Kotlin trim() strips Unicode whitespace, like String.prototype.trim.
  const tokens = raw.trim().match(TOKEN) ?? [];
  return tokens
    .map((token) => {
      if (token.startsWith('"')) return token;
      if (token.endsWith("*")) return token;
      if (containsCjk(token)) return `"${token.replace(/"/g, "")}"`;
      return token;
    })
    .join(" ");
}
